sretoolbox/container/image.py

```diff
@@ -221,13 +268,58 @@ def _get_tags(self):
                 url = next_page["url"]
             else:
                 url = f'{self.registry_api}{next_page["url"]}'
-            response = self._request_get(url)
+            response = self._do_request(url)
 
             tags = response.json()['tags']
             all_tags.extend(tags)
 
         return all_tags
 
+    def _handle_cache_response(self, url):
+        # Uses a dispatch table to handle how different registries handle
+        # responses that are already present in the response cache. We will
+        # favor proper conditional requests if the registry supports it. The
+        # handling methods will update the response cache if needed.
+        return getattr(
+            self, self._HANDLE_RESPONSE_CACHE_METHODS[self.registry]
+        )(url)
+
+    def _handle_conditional_request(self, url):
+        # Handle response cache entries using conditional requests.
+        headers = {}
+        key = self._get_cache_key(url)
+        cached_response = self.response_cache[key]
+
+        etag = cached_response.headers.get("ETag")
+        if etag is not None:
+            headers["If-None-Match"] = etag
+
+        last_mod = cached_response.headers.get("Last-Modified")
+        if last_mod is not None:
+            headers["If-Modified-Since"] = last_mod
+
+        rsp = self._do_request(url, headers=headers)
+
+        if rsp.status_code != HTTPStatus.NOT_MODIFIED:
+            self.response_cache[key] = rsp
+
+    def _handle_docker_content_digest(self, url):
+        # Handle response cache entries using Docker-Content-Digest header.
+        # This method has been inspired by DockerHub, which doesn't support
+        # proper conditional requests but doesn't count HEAD requests towards
+        # quota. See https://docs.docker.com/docker-hub/download-rate-limit/
+        # to have more details.
+        key = self._get_cache_key(url)
+        cached_response = self.response_cache[key]
+        header = "Docker-Content-Digest"
+
+        rsp = self._do_request(url, requests.head)
+        if header in rsp.headers:
+            if rsp.headers.get(header) != cached_response.headers.get(header):
+                self.response_cache[key] = self._do_request(url)
+        else:
+            self.response_cache[key] = self._do_request(url)
```

Comment:
This one also would need some reorg, but it will greatly improve readability, imho...
```
        rsp = self._do_request(url, requests.head)
        if header not in rsp.headers:
            return self._do_request(url)
        if rsp.headers.get(header) != cached_response.headers.get(header):
            return self._do_request(url)
        return cached_response
```

Reply:
Refactored as suggested in https://github.com/app-sre/sretoolbox/commit/21859d80746473245da307a80d9744b040e0590e